src/V3Tristate.cpp

```diff
@@ -715,64 +734,92 @@ class TristateVisitor final : public TristateBaseVisitor {
+    bool isAssignmentNotStrongerThanStrength(AstAssignW* assignp, uint8_t strength) {
+        // If the value of the RHS is known and has all bits equal, only strength corresponding to
+        // its value is taken into account. In opposite case, both strengths are compared.
+        const uint8_t strength0 = getStrength(assignp, 0);
+        const uint8_t strength1 = getStrength(assignp, 1);
+        return (strength0 <= strength && strength1 <= strength)
+               || (strength0 <= strength && assignmentOfValueOnAllBits(assignp, 0))
+               || (strength1 <= strength && assignmentOfValueOnAllBits(assignp, 1));
+    }
+
+    void removeNotStrongerAssignments(Assigns& assigns, AstAssignW* strongestp,
+                                      uint8_t greatestKnownStrength) {
         // Weaker assignments are these assignments that can't change the final value of the net.
-        // If the value of the RHS is known, only strength corresponding to its value is taken into
-        // account. Assignments of constants that have bits both 0 and 1 are skipped here, because
-        // it would involve handling parts of bits separately.
-
-        // First, the strongest assignment, that has value on the RHS consisting of only 1 or only
-        // 0, has to be found.
-        AstAssignW* const strongest0p = getStrongestAssignmentOfValue(assigns, 0);
-        AstAssignW* const strongest1p = getStrongestAssignmentOfValue(assigns, 1);
-        AstAssignW* strongestp = nullptr;
-        uint8_t greatestKnownStrength = 0;
-        const auto getIfStrongest = [&](AstAssignW* const strongestCandidatep, bool value) {
-            if (!strongestCandidatep) return;
-            uint8_t strength = getStrength(strongestCandidatep, value);
-            if (strength >= greatestKnownStrength) {
-                greatestKnownStrength = strength;
-                strongestp = strongestCandidatep;
+        // They can be safely removed. Assignments of the same strength are also removed, because
+        // duplicates aren't needed. One problem is with 2 assignments of different values and
+        // equal strengths. It should result in assignment of x value, but these values aren't
+        // supported now.
+        auto removedIt = std::remove_if(assigns.begin(), assigns.end(), [&](AstAssignW* assignp) {
+            if (assignp == strongestp) return false;
+            if (isAssignmentNotStrongerThanStrength(assignp, greatestKnownStrength)) {
+                // Vertices corresponding to nodes from removed assignment's subtree have to be
+                // removed.
+                m_tgraph.deleteVerticesFromSubtree(assignp);
+                VL_DO_DANGLING(pushDeletep(assignp->unlinkFrBack()), assignp);
+                return true;
+            }
+            return false;
+        });
+        assigns.erase(removedIt, assigns.end());
+    }
+
+    void removeAssignmentsNotStrongerThanUniformConstant() {
```

Comment:
I'm afraid some of these names are a bit too long :) What is the uniform constant?

Reply:
I called it a constant equal to '0 or to '1.
I know that they are long, but I wanted them to be more self-descriptive and I couldn't find anything shorter. But if you have any ideas, please tell me.